pkg/machine/cleanup.go

```diff
@@ -0,0 +1,66 @@
+package machine
+
+import (
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+
+	"github.com/sirupsen/logrus"
+)
+
+type CleanupCallback struct {
+	Funcs []func() error
+	mu    sync.Mutex
+}
+
+func (c *CleanupCallback) Clean(err *error) {
+	// do not remove created files if the init is successful
+	if *err == nil {
+		return
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	for _, cleanfunc := range c.Funcs {
+		if err := cleanfunc(); err != nil {
+			logrus.Error(err)
+		}
+	}
+}
+
+func (c *CleanupCallback) CleanOnSignal(err *error) {
+	ch := make(chan os.Signal, 1)
+	signal.Notify(ch, os.Interrupt, syscall.SIGTERM)
+
+	_, ok := <-ch
+	if !ok {
+		return
+	}
+	// if err exists, then we let the deferred clean function handle the cleanup
+	// this prevents a race where both defer and cleanonsignal tries to handle cleanups
+	if *err != nil {
```

Comment:
There is a race on this particular read because there is no happens-before between the memory write in the machine function and the check here in the signal handler (check out https://go.dev/ref/mem for reasons why). There is also a timing problem where you may check just before it is set. However, since you could leverage your lock below and the callback slice itself to address this issue (see below)